data_spec_validator/spec/wrappers.py

```diff
@@ -14,3 +14,32 @@ def validate(self, value, extra, data):
         msg = ''.join(error.args)
         new_error = type(error)(f'not({msg})')
         return not ok, new_error
+
+class SpecObject:
+    pass
+
+class Spec:
+    __spec_object = None
+
+    @classmethod
+    def __get_reserved_attributes(cls):
+        return dir(Spec)
+
+    @classmethod
+    def __create_specObject(cls):
+        spec = SpecObject()
+        reserved_attributes = cls.__get_reserved_attributes()
+        for attribute in dir(cls):
+            if attribute in reserved_attributes: continue
+
+            setattr(spec, attribute, getattr(cls, attribute))
+
+        return spec
+
+    @classmethod
+    def validate(cls, data, **kwargs):
+        if (cls.__spec_object is None):
+            cls.__spec_object = cls.__create_specObject()
+
+        return validate_data_spec(data, cls.__spec_object, **kwargs)
```

Comment:
Could you please briefly describe the reason why the `SpecObject` type and its instance is required here?

Reply:
That's a good question. Because I added a new validate function. If I inject that cls variable into the validate_data_spec in #44 without new spec instance.  "validate" will be return as a field that needs to be validated by  [_extract_fields](https://github.com/hardcoretech/data-spec-validator/blob/7ee0944ca9899d565ad04ed82ca26bb402970958/data_spec_validator/spec/validators.py#L47), which will cause an error. 

In the first place, I was trying to make some small changes for this repo. So I work around to avoiding modify that _extract_fields function. Overriding` __dir__` is another way has been tried. But it didn't change any result  from `dir(Spec)`, only works for `dir(Spec())`.

In summary, that's why I created a new spec class for validate_data_spec.  The new  instance is because we don't want side effects from other spec object. That can keep it isolated. This way might not be a best solution, any suggestion is welcome! 😃

Reply:
BTW, don't feel bad if you guys want to reject this pull request. this pr more like a  side project. It let me can give it a try to Python.

Reply:
I like the idea `SomeSpec.validate(data, **kwargs)`.

Though there are two concerns to me, the first one is the derived class(and its parents') members (i.e. attributes, properties...) are also listed when a call `dir()` on it.  That's the reason why `def __create_specObject` is here.
The other one is I don't want to create any Spec instance as it should be just a descriptor-like class to indicate what fields to validate.

Instead of calling `dir()`,  it seems valuable enhancing the `_extract_fields` by using `SOME_CLASS.__dict__.keys()` to keep leaf-class-level attributes (parents' attributes are excluded).  
So that the `class Spec` may just have one method `validate`
```
    @classmethod
    def validate(cls, data, **kwargs):
        return validate_data_spec(data, cls, **kwargs)
```

Reply:
OK, what you do here is to make `_extract_fields` work as usual by passing in a subclass without listing fields of the base class `Spec`, right?

I'd suggest to modify `_extract_fields` and make it either
* support both classes: general classes and `Spec` subclasses, or
* check if the class has some specific method, such as `list_fields`.

This will make the code more readable, and you don't need those `setattr`s or any instance at all.

Reply:
hi @kilikkuo and @xeonchen,

Thanks for your suggestions. I figured out the metaclass usage and use it to organize a spec_field. that can be used in _extract_fields function.

About changes in _extract_fields, I thought the condition should be `isinstance(checker, Spec)`, but I can import Spec into validators.py. So I used `hasattr(checker, 'spec_fields')` instead. it could be a potential bug if other general class also has the same name field need to be validated. so I named the spec_fields field as uncommon as possible.

that's pretty much, any comment or suggestions is still welcome.

Reply:
There is another way, all we have to do just only replace one line of code like:
```python
def _extract_fields(checker):
+    return list(filter(lambda f: isinstance(getattr(checker, f), Checker), dir(checker)))
-    return list(filter(lambda f: type(f) == str and not (f.startswith('__') and f.endswith('__')), dir(checker)))
```

that's the simplest approach. we even don't need the SpecMeta metaclass, but it could changed origin behavior. I not sure how you used it in production. this change could by risky. That's the reason why I tired to walk around.